glotaran/analysis/util.py

```diff
@@ -53,66 +62,78 @@ def calculate_matrix(
         dataset_model.swap_dimensions()
 
     for scale, megacomplex in megacomplex_iterator():
-        this_matrix = megacomplex.calculate_matrix(dataset_model, indices)
+        this_clp_labels, this_matrix = megacomplex.calculate_matrix(dataset_model, indices)
 
         if scale is not None:
             this_matrix *= scale
 
         if matrix is None:
+            clp_labels = this_clp_labels
             matrix = this_matrix
         else:
-            matrix, this_matrix = xr.align(matrix, this_matrix, join="outer", copy=False)
-            matrix = matrix.fillna(0)
-            matrix += this_matrix.fillna(0)
+            tmp_clp_labels = clp_labels + [c for c in this_clp_labels if c not in clp_labels]
+            tmp_matrix = np.zeros((matrix.shape[0], len(tmp_clp_labels)), dtype=np.float64)
+            for idx, label in enumerate(tmp_clp_labels):
+                if label in clp_labels:
+                    tmp_matrix[:, idx] += matrix[:, clp_labels.index(label)]
+                if label in this_clp_labels:
+                    tmp_matrix[:, idx] += this_matrix[:, this_clp_labels.index(label)]
+            clp_labels = tmp_clp_labels
+            matrix = tmp_matrix
 
     if global_model:
         dataset_model.swap_dimensions()
 
-    return matrix
+    return CalculatedMatrix(clp_labels, matrix)
+
+
+@nb.jit(nopython=True, parallel=True)
+def apply_weight(matrix, weight):
+    for i in range(matrix.shape[1]):
```

Comment:
```suggestion
    for i in nb.prange(matrix.shape[1]):
```
Not sure if `numba` does that already because of the `parallel=True`, but it is better to make sure of it.